In [378]:
# Import Libraries to read and plot datasets
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale
from collections import Counter
import regex

In [379]:
crash_data = pd.read_csv("hq1d-p-app52dopendataexport2025crashes.csv")

C:\Users\Evan\AppData\Local\Temp\ipykernel_17200\3292712654.py:1: DtypeWarning: Columns (4,12,13,23,24,34,43,53,60,64,65,67,68,69,70,71,72,73) have mixed types. Specify dtype option on import or set low_memory=False.
  crash_data = pd.read_csv("hq1d-p-app52dopendataexport2025crashes.csv")


In [380]:
crash_data.head()

,Collision Id,Report Number,Report Version,Is Preliminary,NCIC Code,Crash Date Time,Crash Time Description,Beat,City Id,City Code,...,IsCountyRoad,IsFreeway,CHP555Version,IsAdditonalObjectStruck,NotificationDate,NotificationTimeDescription,HasDigitalMediaFiles,EvidenceNumber,IsLocationReferToNarrative,IsAOIOneSameAsLocation
0,4550265,9685-2025-00096,1,False,9685,1/18/2025 3:20:00 PM,1520,039,994.0,3300.0,...,False,False,4,NaN,1/18/2025 3:25:00 PM,1525,False,NaN,NaN,True
1,4550264,250110005,1,False,4116,1/10/2025 8:28:00 AM,828,A,1310.0,4116.0,...,NaN,NaN,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,4550263,9580-2025-00214,1,False,9580,1/23/2025 8:45:00 AM,845,171,535.0,1942.0,...,False,True,4,NaN,1/23/2025 8:47:00 AM,0847,False,NaN,NaN,True
3,4550262,9765-2025-00066,1,False,9765,1/15/2025 12:30:00 PM,1230,041,1701.0,5606.0,...,False,True,4,NaN,1/15/2025 12:35:00 PM,1235,True,EMBEDDED,NaN,True
4,4550261,9426-2025-00035,1,False,9426,1/20/2025 12:30:00 PM,1230,460,397.0,1500.0,...,True,False,4,NaN,1/20/2025 12:32:00 PM,1232,False,NaN,NaN,True


In [381]:
crash_data.columns

Index(['Collision Id', 'Report Number', 'Report Version', 'Is Preliminary',
       'NCIC Code', 'Crash Date Time', 'Crash Time Description', 'Beat',
       'City Id', 'City Code', 'City Name', 'County Code', 'City Is Active',
       'City Is Incorporated', 'Collision Type Code',
       'Collision Type Description', 'Collision Type Other Desc',
       'Day Of Week', 'DispatchNotified', 'HasPhotographs', 'HitRun',
       'IsAttachmentsMailed', 'IsDeleted', 'IsHighwayRelated', 'IsTowAway',
       'JudicialDistrict', 'MotorVehicleInvolvedWithCode',
       'MotorVehicleInvolvedWithDesc', 'MotorVehicleInvolvedWithOtherDesc',
       'NumberInjured', 'NumberKilled', 'Weather 1', 'Weather 2',
       'Road Condition 1', 'Road Condition 2', 'Special Condition',
       'LightingCode', 'LightingDescription', 'Latitude', 'Longitude',
       'MilepostDirection', 'MilepostDistance', 'MilepostMarker',
       'MilepostUnitOfMeasure', 'PedestrianActionCode', 'PedestrianActionDesc',
       'PreparedDate',

In [382]:
crash_parties_data = pd.read_csv("a2676918-a825-4b77-8e5c-6eadb38d6b1a.csv")

In [383]:
crash_parties_data.columns

Index(['_id', 'PartyId', 'CollisionId', 'PartyNumber', 'PartyType',
       'IsAtFault', 'IsOnDutyEmergencyVehicle', 'IsHitAndRun', 'AirbagCode',
       'AirbagDescription', 'SafetyEquipmentCode',
       'SafetyEquipmentDescription', 'Special Information',
       'Other Associate Factor', 'Inattention', 'DirectionOfTravel',
       'StreetOrHighwayName', 'SpeedLimit', 'MovementPrecCollCode',
       'MovementPrecCollDescription', 'SobrietyDrugPhysicalCode1',
       'SobrietyDrugPhysicalDescription1', 'SobrietyDrugPhysicalCode2',
       'SobrietyDrugPhysicalDescription2', 'GenderCode', 'GenderDescription',
       'StatedAge', 'DriverLicenseClass', 'DriverLicenseStateCode', 'RaceCode',
       'RaceDesc', 'Vehicle1TypeId', 'Vehicle1TypeDesc', 'Vehicle1Year',
       'Vehicle1Make', 'Vehicle1Model', 'Vehicle1Color', 'V1IsVehicleTowed',
       'Vehicle2TypeId', 'Vehicle2TypeDesc', 'Vehicle2Year', 'Vehicle2Make',
       'Vehicle2Model', 'Vehicle2Color', 'V2IsVehicleTowed', 'Lane',
       'ThruLa

In [384]:
crash_data = crash_data.drop(['Report Number','Report Version','Is Preliminary','NCIC Code','City Is Active',
       'City Is Incorporated', 'Collision Type Code','HasPhotographs', 'IsAttachmentsMailed', 'IsDeleted','JudicialDistrict', 'MotorVehicleInvolvedWithCode', 'MilepostDirection', 'MilepostDistance', 'MilepostMarker',
       'MilepostUnitOfMeasure', 'PreparedDate','ReportingDistrict',
       'ReportingDistrictCode', 'ReviewedDate', 'RoadwaySurfaceCode',
       'SecondaryDirection', 'SecondaryDistance', 'SecondaryRoad',
       'SecondaryUnitOfMeasure', 'SketchDesc', 'TrafficControlDeviceCode',
       'CreatedDate', 'ModifiedDate','CHP555Version', 'NotificationTimeDescription', 'HasDigitalMediaFiles', 'EvidenceNumber',
       'IsLocationReferToNarrative', 'DispatchNotified','Crash Time Description', 'Beat', 'City Id', 'City Code','Weather 2','Road Condition 2','Special Condition','Collision Type Other Desc','NotificationDate', 'IsAOIOneSameAsLocation',
        'IsCountyRoad'], axis = 1)

In [385]:
crash_data.columns

Index(['Collision Id', 'Crash Date Time', 'City Name', 'County Code',
       'Collision Type Description', 'Day Of Week', 'HitRun',
       'IsHighwayRelated', 'IsTowAway', 'MotorVehicleInvolvedWithDesc',
       'MotorVehicleInvolvedWithOtherDesc', 'NumberInjured', 'NumberKilled',
       'Weather 1', 'Road Condition 1', 'LightingCode', 'LightingDescription',
       'Latitude', 'Longitude', 'PedestrianActionCode', 'PedestrianActionDesc',
       'Primary Collision Factor Code', 'Primary Collision Factor Violation',
       'PrimaryCollisionFactorIsCited', 'PrimaryCollisionPartyNumber',
       'PrimaryRoad', 'IsFreeway', 'IsAdditonalObjectStruck'],
      dtype='object')

In [386]:
crash_data_roads = crash_data[['Collision Id','PrimaryRoad','IsFreeway','IsHighwayRelated']]

In [387]:
crash_data_roads.head()

,Collision Id,PrimaryRoad,IsFreeway,IsHighwayRelated
0,4550265,SR-79 N/B,False,True
1,4550264,N EL CAMINO REAL,NaN,False
2,4550263,SR-170 S/B FROM SHERMAN WAY,True,True
3,4550262,SR-126 E/B,True,True
4,4550261,SR-46 W/B,False,False


In [388]:
crash_data_roads = crash_data_roads.fillna(False, axis = 1) #fill in NA fields with False since it should only be True or False

In [389]:
crash_data_roads.count() # confirming that the fields that should 

Collision Id        393510
PrimaryRoad         393510
IsFreeway           393510
IsHighwayRelated    393510
dtype: int64

In [390]:
freeway_crashes = crash_data_roads[crash_data_roads['IsFreeway'] == True]

In [391]:
freeway_crashes

,Collision Id,PrimaryRoad,IsFreeway,IsHighwayRelated
2,4550263,SR-170 S/B FROM SHERMAN WAY,True,True
3,4550262,SR-126 E/B,True,True
5,4550260,SR-170 N/B TO SHERMAN WAY E/B,True,True
6,4550259,SR-126 E/B,True,True
8,4550257,I-980 W/B,True,True
...,...,...,...,...
392786,5107217,SR-238,True,True
392787,5107190,I-5 (SANTA ANA FREEWAY) S/B,True,True
392794,5106858,US-101 N/B,True,True
392799,5106809,I-10 (SAN BERNARDINO FREEWAY) E/B,True,True


In [392]:
local_crashes = crash_data_roads[crash_data_roads['IsFreeway'] == False]

In [393]:
local_crashes

,Collision Id,PrimaryRoad,IsFreeway,IsHighwayRelated
0,4550265,SR-79 N/B,False,True
1,4550264,N EL CAMINO REAL,False,False
4,4550261,SR-46 W/B,False,False
7,4550258,SR-74 (ORTEGA HWY) W/B,False,True
10,4550253,83RD ST,False,False
...,...,...,...,...
393505,5105490,MAIN ST,False,False
393506,5105489,LA VETA AVE,False,False
393507,5105488,CANNON ST,False,False
393508,5105487,CHAPMAN AVE,False,True


In [394]:
assert(len(local_crashes) + len(freeway_crashes)) == 393510 # assert that the divide between local and highway crashes add up to the original amount

In [395]:
local_crashes.to_csv("local_crashes.csv")

In [396]:
freeway_crashes.to_csv("freeway_crashes.csv")

In [397]:
crash_data[crash_data['IsAdditonalObjectStruck'].isna() ==True]['IsAdditonalObjectStruck']

0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
         ... 
393505    NaN
393506    NaN
393507    NaN
393508    NaN
393509    NaN
Name: IsAdditonalObjectStruck, Length: 370507, dtype: object

In [398]:
crash_data[crash_data['IsAdditonalObjectStruck'] ==True]['IsAdditonalObjectStruck']

19        True
46        True
56        True
58        True
62        True
          ... 
392787    True
392798    True
392806    True
392811    True
392815    True
Name: IsAdditonalObjectStruck, Length: 17631, dtype: object

In [399]:
crash_data[crash_data['IsAdditonalObjectStruck'] == False]['IsAdditonalObjectStruck']

49        False
122       False
137       False
249       False
258       False
          ...  
387969    False
388561    False
392354    False
392381    False
392813    False
Name: IsAdditonalObjectStruck, Length: 5372, dtype: object

Given that the majority of the values in IsAdditonalObjectStruck is NaN / Null and this was supposed to be only True or False and no additional information on what additional object means, it will be best to drop this column

In [400]:
crash_data = crash_data.drop(['IsAdditonalObjectStruck'], axis = 1)

In [401]:
crash_data.columns

Index(['Collision Id', 'Crash Date Time', 'City Name', 'County Code',
       'Collision Type Description', 'Day Of Week', 'HitRun',
       'IsHighwayRelated', 'IsTowAway', 'MotorVehicleInvolvedWithDesc',
       'MotorVehicleInvolvedWithOtherDesc', 'NumberInjured', 'NumberKilled',
       'Weather 1', 'Road Condition 1', 'LightingCode', 'LightingDescription',
       'Latitude', 'Longitude', 'PedestrianActionCode', 'PedestrianActionDesc',
       'Primary Collision Factor Code', 'Primary Collision Factor Violation',
       'PrimaryCollisionFactorIsCited', 'PrimaryCollisionPartyNumber',
       'PrimaryRoad', 'IsFreeway'],
      dtype='object')

In [402]:
crash_injuries = crash_data[['Collision Id','IsTowAway','MotorVehicleInvolvedWithDesc','MotorVehicleInvolvedWithOtherDesc','NumberInjured','NumberKilled',
'Primary Collision Factor Code', 'Primary Collision Factor Violation', 'PrimaryCollisionFactorIsCited', 'PrimaryCollisionPartyNumber']]

In [403]:
crash_injuries.head()

,Collision Id,IsTowAway,MotorVehicleInvolvedWithDesc,MotorVehicleInvolvedWithOtherDesc,NumberInjured,NumberKilled,Primary Collision Factor Code,Primary Collision Factor Violation,PrimaryCollisionFactorIsCited,PrimaryCollisionPartyNumber
0,4550265,True,NON-COLLISION,NaN,1.0,0.0,A,VC 22107,False,1.0
1,4550264,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,22350 VC,False,1.0
2,4550263,False,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,VC 22107,False,1.0
3,4550262,False,OTHER OBJECT,METAL OBJECT,0.0,0.0,A,VC 22350,False,1.0
4,4550261,True,OTHER MOTOR VEHICLE,NaN,3.0,0.0,A,VC 22450(a),False,1.0


In [404]:
crash_injuries

,Collision Id,IsTowAway,MotorVehicleInvolvedWithDesc,MotorVehicleInvolvedWithOtherDesc,NumberInjured,NumberKilled,Primary Collision Factor Code,Primary Collision Factor Violation,PrimaryCollisionFactorIsCited,PrimaryCollisionPartyNumber
0,4550265,True,NON-COLLISION,NaN,1.0,0.0,A,VC 22107,False,1.0
1,4550264,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,22350 VC,False,1.0
2,4550263,False,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,VC 22107,False,1.0
3,4550262,False,OTHER OBJECT,METAL OBJECT,0.0,0.0,A,VC 22350,False,1.0
4,4550261,True,OTHER MOTOR VEHICLE,NaN,3.0,0.0,A,VC 22450(a),False,1.0
...,...,...,...,...,...,...,...,...,...,...
393505,5105490,True,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,21801(a),False,1.0
393506,5105489,True,OTHER MOTOR VEHICLE,NaN,2.0,0.0,A,21453(a),False,1.0
393507,5105488,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,21801(a),False,1.0
393508,5105487,True,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,22350,False,1.0


In [405]:
crash_injuries['Primary Collision Factor Code'].value_counts()

Primary Collision Factor Code
A    361426
B     12369
C     10314
D      9137
Name: count, dtype: int64

It seems that for almost 92% of all reported accidents in CA in 2025, is the cause of some vehicle code being broken and not B - Other Improper Driving C - Other Than Driver D - Unknown E - Fell Asleep. Thus We should only be focusing on A 


In [406]:
primary_collision_factor_violation_values_A = crash_injuries[crash_injuries['Primary Collision Factor Code'] == 'A']

In [407]:
primary_collision_factor_violation_values_A

,Collision Id,IsTowAway,MotorVehicleInvolvedWithDesc,MotorVehicleInvolvedWithOtherDesc,NumberInjured,NumberKilled,Primary Collision Factor Code,Primary Collision Factor Violation,PrimaryCollisionFactorIsCited,PrimaryCollisionPartyNumber
0,4550265,True,NON-COLLISION,NaN,1.0,0.0,A,VC 22107,False,1.0
1,4550264,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,22350 VC,False,1.0
2,4550263,False,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,VC 22107,False,1.0
3,4550262,False,OTHER OBJECT,METAL OBJECT,0.0,0.0,A,VC 22350,False,1.0
4,4550261,True,OTHER MOTOR VEHICLE,NaN,3.0,0.0,A,VC 22450(a),False,1.0
...,...,...,...,...,...,...,...,...,...,...
393505,5105490,True,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,21801(a),False,1.0
393506,5105489,True,OTHER MOTOR VEHICLE,NaN,2.0,0.0,A,21453(a),False,1.0
393507,5105488,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,21801(a),False,1.0
393508,5105487,True,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,22350,False,1.0


In [408]:
primary_collision_factor_violation_values_A['Primary Collision Factor Violation'].value_counts()

Primary Collision Factor Violation
VC 22350                           55273
VC 22107                           37072
22350                              33791
22107                              31071
VC 21658(A)                        13432
                                   ...  
21658 (a) VC                           1
VC 21664 ON/OFF RAMP- WRONG WAY        1
21464(B)                               1
21800(g)                               1
14601(1A)                              1
Name: count, Length: 2549, dtype: int64

In [409]:
parsed_primary_collision_factor_violation_values_A = primary_collision_factor_violation_values_A['Primary Collision Factor Violation'].copy()

In [410]:
parsed_primary_collision_factor_violation_values_A = primary_collision_factor_violation_values_A['Primary Collision Factor Violation'].astype(str)

In [411]:
parsed_primary_collision_factor_violation_values_A

0            VC 22107
1            22350 VC
2            VC 22107
3            VC 22350
4         VC 22450(a)
             ...     
393505       21801(a)
393506       21453(a)
393507       21801(a)
393508          22350
393509          22350
Name: Primary Collision Factor Violation, Length: 361426, dtype: object

In [412]:
parsed_primary_collision_factor_violation_values_A = parsed_primary_collision_factor_violation_values_A.str.extract(r'(\d+)') # Series.str.extract(regex'(all decimals repeated)')

In [413]:
parsed_primary_collision_factor_violation_values_A

,0
0,22107
1,22350
2,22107
3,22350
4,22450
...,...
393505,21801
393506,21453
393507,21801
393508,22350


parsed_primary_collision_factor_violation_values_A has all the traffic code violations filtered out. Now we need to find what is the distribution of this.

In [414]:
parsed_primary_collision_factor_violation_values_A.value_counts().head(5)

0    
22350    111508
22107     82769
23152     32700
21658     29842
21453     17256
Name: count, dtype: int64

Looking at the top 5 traffic violations involved in accidents, it seems to show that these 5 violations constitute over 75 percent of all accidents where some ticketable or worse offense occured.

These are what those traffic violations are: 

22350: Speeding

22107: Unsafe Lane Change

23152: DUI

21658: Not staying in Lane

21453: Red Light Violation

In [416]:
type(primary_collision_factor_violation_values_A)

pandas.core.frame.DataFrame

In [418]:
primary_collision_factor_violation_values_A.insert(10, 'parsed_values', parsed_primary_collision_factor_violation_values_A)

In [421]:
primary_collision_factor_violation_values_A.drop('Primary Collision Factor Violation', axis = 1 )

,Collision Id,IsTowAway,MotorVehicleInvolvedWithDesc,MotorVehicleInvolvedWithOtherDesc,NumberInjured,NumberKilled,Primary Collision Factor Code,PrimaryCollisionFactorIsCited,PrimaryCollisionPartyNumber,parsed_values
0,4550265,True,NON-COLLISION,NaN,1.0,0.0,A,False,1.0,22107
1,4550264,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,False,1.0,22350
2,4550263,False,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,False,1.0,22107
3,4550262,False,OTHER OBJECT,METAL OBJECT,0.0,0.0,A,False,1.0,22350
4,4550261,True,OTHER MOTOR VEHICLE,NaN,3.0,0.0,A,False,1.0,22450
...,...,...,...,...,...,...,...,...,...,...
393505,5105490,True,OTHER MOTOR VEHICLE,NaN,0.0,0.0,A,False,1.0,21801
393506,5105489,True,OTHER MOTOR VEHICLE,NaN,2.0,0.0,A,False,1.0,21453
393507,5105488,False,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,False,1.0,21801
393508,5105487,True,OTHER MOTOR VEHICLE,NaN,1.0,0.0,A,False,1.0,22350


In [423]:
primary_collision_factor_violation_values_A.to_csv("Main Collision Factors.csv")